### ISSUE TITLE
ChildMdi has multiple rendering issues when using PerMonitorV2 or PerMonitor

### ISSUE THREAD

**Issue by alanhoman** [dotnet/winforms#3491]
* .NET Core Version: 
3.1.400 and v5.0.0-preview.5

* Have you experienced this same bug with .NET Framework?: 
PerMonitor  using app.manifest works on .NET Framework 4.8.  PerMonitorV2 and PerMonitor do not work in .NET Core 3.1/5.0

**Problem description:**
The following issues are occurring with .NET Core 3.1/5.0 HighDpiMode.PerMonitorV2.  The sample project was based on https://github.com/microsoft/dotnet-samples/tree/master/WinForms-HDPI/PerMonitorAware.  These issues prevent us from being able to move to .NET Core.  One of our largest projects has over 1,000 forms and many have multiple Tabs and ListViews that do not render properly on .NET Core due to PerMonitor and PerMonitorV2 issues.  Currently we run with .NET Framework 4.8 in PerMonitor mode and programmatically adjust the fonts on all child forms when the scaling changes.  This PerMonitor workaround does not work in .NET Core. In all the examples, the 144 Dpi monitor is on the left and the 96 Dpi monitor is on the right. Here are [short video clips of issues 1-6](https://github.com/dotnet/winforms/files/4814338/mp4Files.zip).

**Issue 1**: When moving an MdiParent to a higher DPI monitor the Maximized ChildMdi does not render properly
- Run the project and click MdiParent - Simple Layout
- Maximize SimpleLayout, then drag the MdiParent from a 96 DPI monitor to a 144 DPI monitor (150%).
- The tabControl1 control appears to be anchored beyond the top of the form and overlays all other controls.  The tabContol1 is on the bottom half of the screen and is anchored Top, Bottom, Left, Right.  It should remain on the lower half of the form.
- If the tabControl1 anchoring is changed to Top, Left, this issue does not occur.  On our large project we have dozens of forms that use Top, Bottom, Left, Right anchoring so that it will display as much information as possible to the user depending on their form and screen size.
![Issue 1](https://user-images.githubusercontent.com/2184698/85347187-beff5300-b4ac-11ea-8b62-b873c30ed84f.gif)

**Issue 2**: When moving an MdiParent to a higher DPI monitor the ChildMdi (not maximized) has extra space on the right and bottom, although it does not have the rendering issues when maximized as in Issue 1.
- Run the project and click MdiParent - Simple Layout
- Drag the MdiParent from a 96 DPI monitor to a 144 DPI monitor (150%).
- There is considerable extra space introduced on the right and bottom of SimpleLayout.
![Issue 2](https://user-images.githubusercontent.com/2184698/85347620-f28ead00-b4ad-11ea-87a9-89b3ac18fc70.gif)

**Issue 3**: When moving MdiParent to a higher DPI monitor, opening SimpleLayout, and moving it to a lower DPI monitor, the font size increases.  Moving it back increases it more.
- Run the project and click MdiParent - Simple Layout.
- Move the form to a 144 DPI monitor, close the SimpleLayout and go to Tools -> Open New SimpleLayout.
- Move the MdiParent to the 96 DPI monitor.  The form does not render properly, and the font also increases in size.  
- Move the MdiParent back to the 144 DPI monitor.  The font size increases further.  You will notice that the groupBox1 font size is much larger than when the form was first opened on the 144 Dpi monitor.
![Issue 3](https://user-images.githubusercontent.com/2184698/85347734-4b5e4580-b4ae-11ea-876b-0be9aad3f7f5.gif)

**Issue 4**: When moving SimpleForm to a higher DPI monitor it does not render properly
- Run the project and click Simple Layout
- Drag SimpleLayout from a 96 DPI monitor to a 144 DPI monitor (150%).
- On .NET Core, the ListBox label, TextBox label, and textbox2 are cutoff.  The numericUpDown1, comboBox1, listview1 header, and multiline textbox1 controls are all in a small font (not scaled).  
- On .NET Framework 4.8, the numericUpDown1, comboBox1, listview1 header, and multiline textbox1 controls are all in a small font (not scaled).  The labels are not cutoff as they are in .NET Core.
![Issue 4](https://user-images.githubusercontent.com/2184698/85348934-d7be3780-b4b1-11ea-9884-cf30e213fa10.gif)

**Issue 5**: When increasing then decreasing the font size, the form does not render properly
- Run the project and click MdiParent - Simple Layout.
- Select Tools - > Increase Font Size, then Tools -> Decrease Font Size.
- SimpleLayout does not render properly.  Its rendering issue is like Issue 1.
- If HighDpiMode.PerMonitorV2 is disabled in Program.cs, the issue does not occur.  With HighDpiMode.PerMonitor, the issue still occurs.
- With .NET Framework 4.8 and With PerMonitor set in app.manifest (app.config setting commented out), this issue does not occur. This is how our current app runs.
![Issue 5](https://user-images.githubusercontent.com/2184698/85347833-a7c16500-b4ae-11ea-9397-08b341e6289b.gif)

**Issue 6**: Similar issues occur with HighDpiMode.PerMonitor when moving the MdiParent between monitors with different DPI’s.  After 3-4 moves, the form does not render properly.  For .NET Framework 4.8, this issue does not occur.
![Issue 6](https://user-images.githubusercontent.com/2184698/85350040-e22e0080-b4b4-11ea-9de1-abb2de33fa46.gif)

**Issue 7**: When switching the project from .NET Core 3.1 to NET Core 5.0, the Minimize/Maximize/Close buttons no longer appear on the menu strip when SimpleLayout is maximized.

**Expected behavior:**
The SimpleLayout Form should render properly using PerMonitorV2 on monitors with varying DPI's.

**Minimal repro:**
[.NET Core Sample Project](https://github.com/dotnet/winforms/files/4816186/PerMonitorHDPICore.zip)
[.NET Framework Sample Project](https://github.com/dotnet/winforms/files/4816187/PerMonitorHDPIFramework.zip)

**Comment by RussKie** [comment]
Thank you for the detailed report, it's really appreciated.

There has been some work in this area https://github.com/dotnet/winforms/pull/3394, which will be available in Preview7.

> **Issue 7**: When switching the project from .NET Core 3.1 to NET Core 5.0, the Minimize/Maximize/Close buttons no longer appear on the menu strip when SimpleLayout is maximized.

This is tracked under https://github.com/dotnet/winforms/pull/3461

**Comment by alanhoman** [comment]
Has any progress been made to fix these issues?

**Comment by kirsan31** [comment]
> Has any progress been made to fix these issues?

[**merriemcgaw**](https://github.com/dotnet/winforms/issues/3202#issuecomment-626072014):
> We're working on our longer term roadmap in the coming weeks. I'm starting to see more and more requests for a DPI overhaul from multiple channels, so I'm certain it will be one of the next big things once we get the designer work completed.

So, I think there will be no progress until 5.0 is released and we see some things from 6.0.

**Comment by RussKie** [comment]
Correct, layout engine and high DPI related work are currently slated for .NET 6.0 timeframe.

**Comment by alanhoman** [comment]
Will anyone at Microsoft be using my repos as part of the .NET 6.0 High DPI work?  I would like to know the current status and can provide additional information if needed and am willing to test.   These scenarios are critical in order for our mission critical and very large WinForms to migrate to .NET 6.0

**Comment by RussKie** [comment]
We will be definitely looking at those repros when we start actively working on the layout engine.

**Comment by merriemcgaw** [comment]
@dreddy-work is starting the investigation work. It may take a little bit to get our end to end plan sorted, but I wanted to let you know that we're actively working on it now.

**Comment by alanhoman** [comment]
Hi Merrie,

That is great to hear.  If you need a volunteer to test, I am available.

Thanks,
Alan

**Comment by alanhoman** [comment]
Hi @dreddy-work, thanks for the excellent work on the High DPI improvements.  I have the winforms project building and running, but have a very important issue that needs to be resolved.  When I try to use the winforms binaries on one of my existing projects following the steps in https://github.com/dotnet/winforms/blob/main/docs/debugging.md, I cannot get it to work.  **This is critical to the testing I am doing.**

When I try method 1, Copy your changes into the SDK, or method 2, my project won't build and I get the following error: 

Error	CS1705	Assembly 'System.Windows.Forms' with identity 'System.Windows.Forms, Version=42.42.42.42, Culture=neutral, PublicKeyToken=b77a5c561934e089' uses 'System.Windows.Forms.Primitives, Version=42.42.42.42, Culture=neutral, PublicKeyToken=b77a5c561934e089' which has a higher version than referenced assembly 'System.Windows.Forms.Primitives' with identity 'System.Windows.Forms.Primitives, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'	PerMonitorHDPICore	C:\Users\ahoman\source\repos\PerMonitorHDPICore - NET 6.0\PerMonitorHDPCore\CSC	1	N/A

I checked System.Windows.Forms.dll and System.Windows.Forms.Primitives.dll and they both have File version: **42.42.42.42424**.  I am not sure why this is or how to fix it.  I did a clean solution and deleted obj but still have this issue.

Can you please help me?

**Comment by dreddy-work** [comment]
@alanhoman , it seems you copied only System.Windows.Form.dll. But it depends on primitive as well. Please also copy System.Windows.Form. primitive.dll as well. It should unblock.

**Comment by alanhoman** [comment]
I did copy System.Windows.Forms.Primitives.dll to C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App\6.0.0-preview.4.21254.5.  Are these dll's supposed to have File version: 42.42.42.42424?

**Comment by dreddy-work** [comment]
All our test builds carry that version.  If you copied both binaries to same location and  project targeting that 6.0, it should work. Can you share a GIF with copying binaries and running app where you still get this error?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by alanhoman** [comment]
Success!  Thanks for the additional info.  The one out of date dll was C:\Program Files\dotnet\packs\Microsoft.WindowsDesktop.App.Ref\6.0.0-preview.4.21254.5\ref\net6.0\System.Windows.Forms.Primitives.dll.  It wasn't in my ref folder, so I copied it from C:[repo]\winforms\artifacts\bin\System.Windows.Forms\Debug\net6.0.  Once I updated this, the build worked.  

Thanks for the help!

**Comment by dreddy-work** [comment]
Good to hear. And thank you @alanhoman for testing it out. As a note, work on HDPI is still in progress and we have identified some control specific issues ( you will see issues being created for them in next couple of days).  Please let us know how your testing results and issues you still see.

**Comment by alanhoman** [comment]
Great, I am starting to test with the understanding that you are actively making changes.  Should I run off of the dev/dreddy/ContainerControlScalingInHighDpi branch to get your latest changes, the main branch, or some other branch? 

Also when I run using your branch or main, I get the error below when I try to run the original test repo from this issue and select the MdiParent - Simple layout.  Somehow Owner became null after the null check.  Is there something else I need to pull into the repository?
![image](https://user-images.githubusercontent.com/2184698/121401769-fbba9900-c90d-11eb-98d3-cace7abbe076.png)

**Comment by dreddy-work** [comment]
I would suggest to use 'main' branch for any of your testing. This particular issue need to be investigated on why. it doesn't seem to be related to the High Dpi changes. Can you open a new issue for this to investigate further ?

**Comment by alanhoman** [comment]
I have done some initial testing of the High DPI changes and here is what I have found.

1. For the MdiParent form if tabPage1 is the visible tab at the time you move from a 100% to a 150% monitor, then tabPage2 does not layout properly on 150%.  Conversely, if tabPage2 was visible, then after the move tabPage1 will not layout properly.  Also on the 150% monitor, half of the child form header/title bar is visible when the child form is maximized and it should not be visible. 

2. For the Simple Form and MdiParent form, if you start the app on the 150% monitor, it comes up much smaller than if you had started on a 100% monitor and then moved it to a 150% monitor.  If you move it to a 100% and then back to 150%, it is then sized as expected. The expected behavior is that when you start the app on a 150% monitor, it will scale up.

3. For Simple Layout with no default font set in Main() going from 100% to 150% does not layout properly.  If a default font is set, it lays out properly.

**Comment by dreddy-work** [comment]
Thanks @alanhoman for validation. I will start looking into the scenarios you listed here.

**Comment by Jenna-Zheng** [comment]
Verified this issue with Core 6.0.100-preview.6.21323.12, Most of the issues have been fixed, but found several new issues that have been submitted for tracking([#5098](https://github.com/dotnet/winforms/issues/5098), [#5106](https://github.com/dotnet/winforms/issues/5106), [#5105](https://github.com/dotnet/winforms/issues/5105), [#5102](https://github.com/dotnet/winforms/issues/5102))

**Test Project:** [PerMonitorHDPCore.zip](https://github.com/dotnet/winforms/files/6707399/PerMonitorHDPCore.zip)

**Comment by alanhoman** [comment]
I am trying to test all of the High DPI changes from the main branch and recently I started getting an exception that prevents me from doing any testing.  Did something change in the build process or is something broken in the main branch?  I have 6.0.0-preview.5.21301.4 installed, and then copy the latest WinForms dlls from the build into it.

![image](https://user-images.githubusercontent.com/2184698/123516905-5fa1c900-d653-11eb-8850-90b713dce581.png)
Here is the stack trace:
[StackTrace.txt](https://github.com/dotnet/winforms/files/6720279/StackTrace.txt).

**Comment by dreddy-work** [comment]
@alanhoman ,  main branch is constantly updated for latest dependencies. It seems you have couple of build old runtime than that our 'main' branch targets.  Try installing latest runtime.

**Comment by alanhoman** [comment]
I am using the publicly available .NET 6.0 Preview 5 SDK that was released on 6/17/2021.  It looks like winforms is using the yet to be released Preview 7.  Should I be using this?  Do you have a link that describes how to do this since it doesn't look like there is an installer for it yet?  What process do you follow to keep your SDK up-to-date with the winforms work?

**Comment by RussKie** [comment]
We're building against .NET nightlies (which are [automatically updated](https://github.com/dotnet/winforms/pulls?q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+author%3Aapp%2Fdotnet-maestro)).
There are few ways you can go about it - either install the latest nightlies globally (see: https://github.com/dotnet/installer), or prepend the path to .\winforms\\.dotnet to your global %PATH% so your VS and your apps know where the latest require .NET SDK is.

**Comment by alanhoman** [comment]
@RussKie , I installed the dotnet-sdk-6.0.100-preview.7.21326.2 SDK daily build from https://github.com/dotnet/installer/blob/main and everything is working now.  Thanks for the guidance! I will continue testing and open a new issue for each scenario if I find anything.
